### ISSUE TITLE
[DEV] Rework Inventory

### ISSUE THREAD

**Issue by oxedions** [bluebanquise/bluebanquise#709]
- [ ] Merge all inventory global settings into a single commented file with prefixed bb_ variables
- [ ] #705 , BlueBanquise should be usable as a stack or a collection
- [ ] Nodes should be defined without mg and equipment group in yaml files, and then added into mg and equipment groups in separate Ansible syntax files. This will allow nodes of different purposes but with same equipment profiles to be defined in the same equipment group. 
- [ ] Issue #639 has added a better handling for authentication. This should be combined with a new variable like “ep_power_protocol” that specify which protocol in the list can be used to power on/off nodes. 
- [ ] Networks file is not really readable by users. Also, services ip do not easily allow today to bind to an hostname (custom or real server). This could be improved. 
  - [ ] Allow hostnames as service ip
  - [ ] Allow list of ip for service ip (roundrobin, HA, etc)
- [ ] Networks definitions currently require subnet, prefix, broadcast, netmask, etc. This could be reduced to something like `http://10.0.0.0/16` and using python addr filters to generate other values. 
- [ ] General_settings files are a bit messy, would worth improving that. 
- [ ] Equipment_profiles could be split into separated files for easier understanding: ipxe, os, authentication, etc. 
- [ ] External file is not clear at all and should be reworked (or is it really necessary, as it could be just roles default/optional variables?)
- [ ] Move authentication_ variables to ep_, and rename root_password
- [x] https://github.com/bluebanquise/bluebanquise/issues/672
- [ ] Ability to centralize subnet unique values for all nodes (ex: subnet_prefix: 10.10. and each node would have : ip4: .3.1, ip4: .3.2, etc) (to be investigated, heavy, sed is enough)
- [ ] ip4 ability to be dhcp (and not static ip)
- [ ] Ability to not having to define iceberg number in networks (ice1-1 to ice[0-9][a-z])

**Comment by ginomcevoy** [comment]
Some ideas I have collected while using and supporting BB:

- networks.yml: extract the services from this file into another file services.yml. Then refer to the services of a network like this: "{{ services.<iceberg>.<network> }}" or maybe just "{{ services.<network>}}". Users may change services.yml with a higher frequency than they would change networks.yml.
- Have a way to add FQDN entries to the current domain. There is a requirement to use a FQDN (instead of just an IP address) when dealing with virtual IP address (High Availability / load balancing) assigned to a service. Currently, the workaround is to use "external_hosts" but the domain name needs to be hard-coded in the key of the entry. This requirement may be satisfied with "Allow hostnames as service ip", not sure.
- Regarding Feature Request #639 , I think we have a requirement to support Redfish authentication. This should not be too difficult to do in conman role.
- Question related to the item above, are there any plans to support "powerman" over "conman" in core Bluebanquise? I ask because I believe conman development looks stagnant, while powerman looks active and appears to support more protocols.

**Comment by oxedions** [comment]
Dear @ginomcevoy 
Thanks for the proposals 😊

**Comment by oxedions** [comment]
- [ ] Possibility to create nodeset range of nodes in yaml
```
all:
  hosts:
    c[001:100]:
      network_interfaces:
        - interface: eno1
          ip4: 10.10.0.[1:100]
          mac:
            - AA:BB:CC:DD:EE:01
            - AA:BB:CC:DD:EE:02
            - ...
          network: ice1-1
```

**Comment by wlln** [comment]
## Role variables

IMHO, roles should not use variables they don't know anything about.

For instance, let's say `dhcp_server` requires a `dhcp_server_time_servers` variable.
The role is to be applied to a group named `foobar`.
In `inventory/group_vars/foobar/dhcp_server/main.yaml`, `dhcp_server_time_servers` is set:
```yaml
---
dhcp_server_time_servers:
  - 10.0.0.1
  - 10.0.0.2
...
```
The *problem* is that every role would require `role_prefix_variable_name` variables.
What if X roles require the same data ? Having to keep every files in sync is a no-go.

But ...

What if every BB variables are prefixed with `bb_*`? then `bb_time_servers` could be set either manually with:
```yaml
---
bb_time_servers:
  - 10.0.0.1
  - 10.0.02
...
```
of with a jinja2 filter:
```yaml
---
bb_time_servers: "{{ a_super_duper_filter_that_returns_a_list_of_ips }}"
...
```
Now, every role:
  * isn't required to know about some dark variables that could be set or not,
  * can check types, values, regexes, etc, of every `role_prefix_variable_name` variables,
  * can be used as a standalone, using `role_prefix_variable_name` variables,
  * can be used with BlueBanquise, by checking the `bb_that_variable_i_should_use` variable.

## BlueBanquise variables

Now, every role variable is set either in `inventory/{group,host}_vars/{group,host}/role_name/main.yaml` or using `bb_*` variables.
But where are the variables defined? Easy:
  * `inventory/group_vars/all/bluebanquise/main.yaml` to target every group and every host,
  * `inventory/group_vars/group_name/bluebanquise/main.yaml` to target a specific group,
  * `inventory/host_vars/host_name/bluebanquise/main.yaml` to target a specific host,
  * eventually, `playbook/playbook_name.yaml`, though it' might lead to strange behaviors if two playbooks use the same variable with different values.

**Comment by wlln** [comment]
## Role behaviour

All variables must have default values set in `defaults/main.yaml`.

When no `role_prefix_*` and/or no `bb_*` is set, including a role into a playbook should not make that playbook crash because of undefined variables. Instead, it should:
  * warn about theses variables and skip:
```yaml
---
# main.yaml
- name: include_tasks | manage local groups
  ansible.builtin.include_tasks: manage_group.yaml
  when: |
    local_group_group is defined and
    local_group_group is mapping  
...
---
# manage_group.yaml
- name: group | manage local groups
  ansible.builtin.group:
    name: "{{ group }}"
    gid: "{{ local_group_group[group].gid | default(omit) }}"
    state: "{{ local_group_group[group].state | default('present') }}"
    system: "{{ local_group_group[group].system | default(false) }}"
  become: true
  tags:
    - group
  loop_control:
    loop_var: group
  loop: "{{ local_group_group | flatten | sort }}"
  when: |
    group is string and
    (group |
     regex_search('^[a-z_][a-z0-9_-]*[$]?') == group) 
...
```
  * use a fallback variable that will **always** exist:
```yaml
---
- name: assert | check set_hostname_hostname type and value
  ansible.builtin.assert:
    that: |
      (set_hostname_hostname is string) and
      (set_hostname_hostname |
       regex_search('^[A-Za-z][-A-Za-z0-9]{0,63}') == set_hostname_hostname)
  when: set_hostname_hostname is defined

- name: hostname | set hostname
  ansible.builtin.hostname:
    name: "{{ set_hostname_hostname | default(inventory_hostname) }}"
    use: systemd
  become: true
...
```

## Misc

  * [Pre-commit](https://pre-commit.com/) with hooks such as:
    * trailing-whitespace
    * end-of-file-fixer
    * fix-byte-order-marker
    * ansible-lint
    * gitlint
  * roles in separate repositories, so that then can be included in BB using `git submodules`

**Comment by oxedions** [comment]
Dear @wlln 
Thanks a lot for your ideas ! 😊

**Comment by wlln** [comment]
# Bootstraping BB

Until BB 1.5, after a fresh install, BB required a manual configuration part, to set up a few things before running playbooks/roles.

The `bootstrap.sh` script simplify the process and that's awesome !

But I feel like bootstraping BB would become way more convinient if it could be bootstraped from a fresh install just by installing a RPM, a DEB (or even from a [PKGBUILD](https://wiki.archlinux.org/title/PKGBUILD) in a near future :roll_eyes:) and running everything from Ansible, not matter what role are added to the playbook. 

After a fresh install of a management node, you could have a single playbook to deploy a complete cluster. Here is a snippet of hat I have in mind:

```yaml
# playbook/cluster.yaml
# This is the only playbook I should call, ever
- ansible.builtin.import_playbook: boostrap_bluebanquise.yaml # prepare the first management node to deploy the other. Should only be run once, so it requires something like checking that a file exists in `/var/lib/bluebanquise'
- ansible.builtin.import_playbook: configure_management_nodes.yaml # should work with 1 to N management nodes, with or without HA and/or icebergs
- ansible.builtint.import_playbook: configure_switches.yaml # configure every switches with ansible roles/modules. otherwise, should be a manual task, or done before starting the management nodes
- ansible.builtin.import_playbook: configure_storage_nodes.yaml # Should deploy storage nodes, for instance, MinIO, BeeGFS or any distributed storage that can be installed/configured/started with ansible roles/modules
- ansible.builtin.import_playbook: configure_nodes.yaml # should start the compute nodes, N by N, and apply playbook when they are *freshly installed*
# and so on so forth
```

**Comment by oxedions** [comment]
New data model is now converged. Most of these idea were merged into BB 2.X :blush: 
We can close this issue.
